lib/msal-common/src/network/ThrottlingUtils.ts

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { NetworkResponse } from "./NetworkManager";
+import { ServerAuthorizationTokenResponse } from "../server/ServerAuthorizationTokenResponse";
+import { Constants, HeaderNames } from "../utils/Constants";
+import { CacheManager } from "../cache/CacheManager";
+import { StringUtils } from "../utils/StringUtils";
+
+export type RequestThumbprint = {
+    clientId: string;
+    authority: string;
+    scopes: Array<string>;
+    homeAccountIdentifier?: string;
+}
+
+export type RequestThumbprintValue = {
+    // Unix-time value representing the expiration of the throttle
+    throttleTime: number;
+    // Information provided by the server
+    error?: string;
+    errorCodes?: Array<string>;
+    errorMessage?: string;
+    subError?: string;
+}
+
+export class ThrottlingUtils {
+
+    static generateThrottlingStorageKey(thumbprint: RequestThumbprint | RequestThumbprintValue): string {
+        return `${Constants.THROTTLE_PREFIX}.${JSON.stringify(thumbprint)}`;
+    }
+
+    static preProcess(cacheManager: CacheManager, thumbprint: RequestThumbprint): void {
+        const key = ThrottlingUtils.generateThrottlingStorageKey(thumbprint);
+        const storageValue = cacheManager.getItem(key) as string;
+
+        if (storageValue) {
+            const parsedValue = StringUtils.jsonParseHelper(storageValue);
+
+            if (parsedValue.throttleTime >= Date.now()) {
```

Comment:
To account for possibility of `jsonParseHelper` failing:
```suggestion
            if (parsedValue && parsedValue.throttleTime >= Date.now()) {
```

Reply:
Done